cmd/snap-confine/mount-support.c

```diff
@@ -319,6 +320,175 @@ static void sc_do_mounts(const char *scratch_dir,
+ */
+static void sc_replicate_base_rootfs(const char *scratch_dir,
+                                     const char *rootfs_dir,
+                                     const struct sc_mount *root_mounts)
+{
+	// First of all, fix the root filesystem:
+	// - remove write permissions for group and others
+	// - set the owner to root:root
+	if (chmod(scratch_dir, 0755) < 0) {
+		die("cannot change permissions on \"%s\"", scratch_dir);
+	}
+	if (chown(scratch_dir, 0, 0) < 0) {
+		die("cannot change ownership on \"%s\"", scratch_dir);
+	}
+
+	int rootfs_fd SC_CLEANUP(sc_cleanup_close) = -1;
+	// Note that the rootfs here is a path like /snap/<snap>/current, which is
+	// always a symbolic link. Therefore, we cannot use O_NOFOLLOW here.
+	rootfs_fd = open(rootfs_dir, O_RDONLY | O_DIRECTORY | O_CLOEXEC);
+	if (rootfs_fd < 0) {
+		die("cannot open directory \"%s\"", rootfs_dir);
+	}
+
+	DIR *rootfs SC_CLEANUP(sc_cleanup_closedir) = fdopendir(rootfs_fd);
+	if (rootfs == NULL) {
+		die("cannot open directory \"%s\" from file descriptor", rootfs_dir);
+	}
+
+	// Will create folders/links as 0:0
+	sc_identity old = sc_set_effective_identity(sc_root_group_identity());
+
+	char full_path[PATH_MAX];
+	// After we construct each entry's full path, we'll need to obtain the
+	// entry's absolute path in the new rootfs, that is with the `scratch_dir`
+	// prefix removed (the path_in_rootfs variable below). We'll do this by
+	// computing the length of the `scratch_dir` prefix now and then using it
+	// as the offset in `full_path` where the '/' of the confined silesystem is
+	// located.
+	const size_t scratch_dir_length = strlen(scratch_dir);
+
+	while (true) {
+		errno = 0;
+		struct dirent *ent = readdir(rootfs);
+		if (ent == NULL) break;
+
+		if (sc_streq(ent->d_name, ".") || sc_streq(ent->d_name, "..")) {
+			continue;
+		}
+
+		sc_must_snprintf(full_path, sizeof(full_path), "%s/%s",
+		                 scratch_dir, ent->d_name);
+		if (ent->d_type == DT_DIR) {
+			if (mkdir(full_path, 0755) < 0) {
+				die("cannot create directory \"%s\"", full_path);
+			}
+
+			// If the directory is listed in root_mounts skip it,
+			// as it will be created and mounted in
+			// sc_bootstrap_mount_namespace() later.
+			bool skip_dir = false;
+			const char *path_in_rootfs = full_path + scratch_dir_length;
+			for (const struct sc_mount *mnt = root_mounts;
+			     mnt->path != NULL; mnt++) {
+				if (sc_streq(path_in_rootfs, mnt->path) ||
+				    sc_streq(path_in_rootfs, mnt->altpath)) {
+					skip_dir = true;
+					break;
+				}
+			}
+			if (skip_dir) {
+				continue;
+			}
+			// Also skip the /snap directory, as we'll mount it later
+			if (sc_streq(path_in_rootfs, "/snap")) {
+				continue;
+			}
+
+			char src_path[PATH_MAX];
+			sc_must_snprintf(src_path, sizeof(src_path), "%s/%s",
+					 rootfs_dir, ent->d_name);
+			sc_do_mount(src_path, full_path, NULL, MS_REC | MS_BIND, NULL);
+			sc_do_mount("none", full_path, NULL, MS_REC | MS_SLAVE, NULL);
```

Comment:
could you add a comment about why we use MS_SLAVE, as the original bind mount below didn't have that

Reply:
I added a commit removing it, since we are doing a recursive remount with MS_SLAVE a bit later, on the scratch directory.